File: merger/asyncapi.go
package merger

import (
	"errors"
	"fmt"
	"sort"

	"github.com/xraph/farp"
)

// AsyncAPISpec represents a simplified AsyncAPI 2.x/3.x specification.
type AsyncAPISpec struct {
	AsyncAPI   string                 `json:"asyncapi"`
	Info       Info                   `json:"info"`
	Servers    map[string]AsyncServer `json:"servers,omitempty"`
	Channels   map[string]Channel     `json:"channels"`
	Components *AsyncComponents       `json:"components,omitempty"`
	Security   []map[string][]string  `json:"security,omitempty"`
	Extensions map[string]any         `json:"-"`
}

// AsyncServer represents an AsyncAPI server (broker connection).
type AsyncServer struct {
	URL         string         `json:"url"`
	Protocol    string         `json:"protocol"` // kafka, amqp, mqtt, ws, etc.
	Description string         `json:"description,omitempty"`
	Variables   map[string]any `json:"variables,omitempty"`
	Bindings    map[string]any `json:"bindings,omitempty"`
}

// Channel represents an AsyncAPI channel.
type Channel struct {
	Description string         `json:"description,omitempty"`
	Subscribe   *Operation     `json:"subscribe,omitempty"`
	Publish     *Operation     `json:"publish,omitempty"`
	Parameters  map[string]any `json:"parameters,omitempty"`
	Bindings    map[string]any `json:"bindings,omitempty"`
	Extensions  map[string]any `json:"-"`
}

// AsyncComponents represents AsyncAPI components.
type AsyncComponents struct {
	Messages        map[string]map[string]any      `json:"messages,omitempty"`
	Schemas         map[string]map[string]any      `json:"schemas,omitempty"`
	SecuritySchemes map[string]AsyncSecurityScheme `json:"securitySchemes,omitempty"`
}

// AsyncSecurityScheme represents an AsyncAPI security scheme.
type AsyncSecurityScheme struct {
	Type             string         `json:"type"` // userPassword, apiKey, X509, symmetricEncryption, asymmetricEncryption, httpApiKey, http, oauth2, openIdConnect
	Description      string         `json:"description,omitempty"`
	Name             string         `json:"name,omitempty"`             // For apiKey, httpApiKey
	In               string         `json:"in,omitempty"`               // For apiKey, httpApiKey: user, password, query, header, cookie
	Scheme           string         `json:"scheme,omitempty"`           // For http
	BearerFormat     string         `json:"bearerFormat,omitempty"`     // For http bearer
	OpenIdConnectURL string         `json:"openIdConnectUrl,omitempty"` // For openIdConnect
	Flows            map[string]any `json:"flows,omitempty"`            // For oauth2
}

// AsyncAPIServiceSchema wraps an AsyncAPI schema with its service context.
type AsyncAPIServiceSchema struct {
	Manifest *farp.SchemaManifest
	Schema   any
	Parsed   *AsyncAPISpec
}

// AsyncAPIMerger handles AsyncAPI schema composition.
type AsyncAPIMerger struct {
	config MergerConfig
}

// NewAsyncAPIMerger creates a new AsyncAPI merger.
func NewAsyncAPIMerger(config MergerConfig) *AsyncAPIMerger {
	return &AsyncAPIMerger{
		config: config,
	}
}

// AsyncAPIMergeResult contains the merged AsyncAPI spec and metadata.
type AsyncAPIMergeResult struct {
	Spec             *AsyncAPISpec
	IncludedServices []string
	ExcludedServices []string
	Conflicts        []Conflict
	Warnings         []string
}

// MergeAsyncAPI merges multiple AsyncAPI schemas from service manifests.
func (m *AsyncAPIMerger) MergeAsyncAPI(schemas []AsyncAPIServiceSchema) (*AsyncAPIMergeResult, error) {
	result := &AsyncAPIMergeResult{
		Spec: &AsyncAPISpec{
			AsyncAPI: "2.6.0", // Use AsyncAPI 2.6
			Info: Info{
				Title:       m.config.MergedTitle,
				Description: m.config.MergedDescription,
				Version:     m.config.MergedVersion,
			},
			Servers:  make(map[string]AsyncServer),
			Channels: make(map[string]Channel),
			Components: &AsyncComponents{
				Messages: make(map[string]map[string]any),
				Schemas:  make(map[string]map[string]any),
			},
			Extensions: make(map[string]any),
		},
		IncludedServices: []string{},
		ExcludedServices: []string{},
		Conflicts:        []Conflict{},
		Warnings:         []string{},
	}

	// Track what we've seen for conflict detection
	seenChannels := make(map[string]string)        // channel -> service
	seenMessages := make(map[string]string)        // message -> service
	seenServers := make(map[string]string)         // server -> service
	seenSecuritySchemes := make(map[string]string) // security scheme -> service

	// Process each schema
	for _, schema := range schemas {
		serviceName := schema.Manifest.ServiceName

		// Check if this schema should be included
		if !shouldIncludeAsyncAPIInMerge(schema) {
			result.ExcludedServices = append(result.ExcludedServices, serviceName)

			continue
		}

		result.IncludedServices = append(result.IncludedServices, serviceName)

		// Parse the schema if not already parsed
		if schema.Parsed == nil {
			parsed, err := ParseAsyncAPISchema(schema.Schema)
			if err != nil {
				result.Warnings = append(result.Warnings,
					fmt.Sprintf("Failed to parse AsyncAPI schema for %s: %v", serviceName, err))

				continue
			}

			schema.Parsed = parsed
		}

		// Get composition config
		compConfig := getAsyncAPICompositionConfig(schema.Manifest)
		strategy := m.getConflictStrategy(compConfig)

		// Determine prefixes
		channelPrefix := getAsyncAPIChannelPrefix(schema.Manifest, compConfig)
		messagePrefix := getAsyncAPIMessagePrefix(schema.Manifest, compConfig)

		// Merge channels
		for channelName, channel := range schema.Parsed.Channels {
			prefixedName := channelName
			if channelPrefix != "" {
				prefixedName = channelPrefix + "." + channelName
			}

			// Check for channel conflicts
			if existingService, exists := seenChannels[prefixedName]; exists {
				conflict := Conflict{
					Type:     "channel",
					Item:     channelName,
					Services: []string{existingService, serviceName},
					Strategy: strategy,
				}

				switch strategy {
				case farp.ConflictStrategyError:
					return nil, fmt.Errorf("channel conflict: %s exists in both %s and %s",
						channelName, existingService, serviceName)

				case farp.ConflictStrategySkip:
					conflict.Resolution = "Skipped channel from " + serviceName
					result.Conflicts = append(result.Conflicts, conflict)

					continue

				case farp.ConflictStrategyOverwrite:
					conflict.Resolution = fmt.Sprintf("Overwritten with %s version", serviceName)
					result.Conflicts = append(result.Conflicts, conflict)

				case farp.ConflictStrategyPrefix:
					prefixedName = serviceName + "." + channelName
					conflict.Resolution = "Prefixed to " + prefixedName
					result.Conflicts = append(result.Conflicts, conflict)

				case farp.ConflictStrategyMerge:
					// Merge operations
					channel = mergeChannels(result.Spec.Channels[prefixedName], channel)
					conflict.Resolution = "Merged operations"
					result.Conflicts = append(result.Conflicts, conflict)
				}
			}

			result.Spec.Channels[prefixedName] = channel
			seenChannels[prefixedName] = serviceName
		}

		// Merge components
		if schema.Parsed.Components != nil {
			// Merge messages
			for name, message := range schema.Parsed.Components.Messages {
				prefixedName := name
				if messagePrefix != "" {
					prefixedName = messagePrefix + "_" + name
				}

				if existingService, exists := seenMessages[prefixedName]; exists {
					if strategy == farp.ConflictStrategySkip {
						result.Conflicts = append(result.Conflicts, Conflict{
							Type:       ConflictTypeComponent,
							Item:       name,
							Services:   []string{existingService, serviceName},
							Resolution: "Skipped message from " + serviceName,
							Strategy:   strategy,
						})

						continue
					}
				}

				result.Spec.Components.Messages[prefixedName] = message
				seenMessages[prefixedName] = serviceName
			}

			// Merge schemas
			for name, schemaObj := range schema.Parsed.Components.Schemas {
				prefixedName := messagePrefix + "_" + name
				result.Spec.Components.Schemas[prefixedName] = schemaObj
			}

			// Merge security schemes
			for name, secScheme := range schema.Parsed.Components.SecuritySchemes {
				if existingService, exists := seenSecuritySchemes[name]; exists {
					conflict := Conflict{
						Type:     ConflictTypeSecurityScheme,
						Item:     name,
						Services: []string{existingService, serviceName},
						Strategy: strategy,
					}

					switch strategy {
					case farp.ConflictStrategyError:
						return nil, fmt.Errorf("security scheme conflict: %s exists in both %s and %s",
							name, existingService, serviceName)

					case farp.ConflictStrategySkip:
						conflict.Resolution = "Skipped security scheme from " + serviceName
						result.Conflicts = append(result.Conflicts, conflict)

						continue

					case farp.ConflictStrategyOverwrite:
						conflict.Resolution = fmt.Sprintf("Overwritten with %s version", serviceName)
						result.Conflicts = append(result.Conflicts, conflict)

					case farp.ConflictStrategyPrefix:
						prefixedName := serviceName + "_" + name
						conflict.Resolution = "Prefixed to " + prefixedName
						result.Conflicts = append(result.Conflicts, conflict)
						result.Spec.Components.SecuritySchemes[prefixedName] = secScheme
						seenSecuritySchemes[prefixedName] = serviceName

						continue

					case farp.ConflictStrategyMerge:
						conflict.Resolution = fmt.Sprintf("Merged (overwritten) with %s version", serviceName)
						result.Conflicts = append(result.Conflicts, conflict)
					}
				}

				result.Spec.Components.SecuritySchemes[name] = secScheme
				seenSecuritySchemes[name] = serviceName
			}
		}

		// Merge servers
		for serverName, server := range schema.Parsed.Servers {
			prefixedName := serviceName + "_" + serverName
			if existingService, exists := seenServers[prefixedName]; exists {
				result.Warnings = append(result.Warnings,
					fmt.Sprintf("Server %s from %s overwrites %s", serverName, serviceName, existingService))
			}

			result.Spec.Servers[prefixedName] = server
			seenServers[prefixedName] = serviceName
		}
	}

	return result, nil
}

// ParseAsyncAPISchema parses a raw AsyncAPI schema into structured format.
func ParseAsyncAPISchema(raw any) (*AsyncAPISpec, error) {
	schemaMap, ok := raw.(map[string]any)
	if !ok {
		return nil, errors.New("schema must be a map")
	}

	spec := &AsyncAPISpec{
		Servers:    make(map[string]AsyncServer),
		Channels:   make(map[string]Channel),
		Extensions: make(map[string]any),
	}

	// Parse AsyncAPI version
	if v, ok := schemaMap["asyncapi"].(string); ok {
		spec.AsyncAPI = v
	} else {
		return nil, errors.New("missing asyncapi version")
	}

	// Parse info
	if info, ok := schemaMap["info"].(map[string]any); ok {
		spec.Info = parseInfo(info)
	}

	// Parse servers
	if servers, ok := schemaMap["servers"].(map[string]any); ok {
		spec.Servers = parseAsyncServers(servers)
	}

	// Parse channels
	if channels, ok := schemaMap["channels"].(map[string]any); ok {
		spec.Channels = parseChannels(channels)
	}

	// Parse components
	if components, ok := schemaMap["components"].(map[string]any); ok {
		spec.Components = parseAsyncComponents(components)
	}

	return spec, nil
}

func parseAsyncServers(servers map[string]any) map[string]AsyncServer {
	result := make(map[string]AsyncServer)

	for name, s := range servers {
		if serverMap, ok := s.(map[string]any); ok {
			server := AsyncServer{}
			if url, ok := serverMap["url"].(string); ok {
				server.URL = url
			}

			if protocol, ok := serverMap["protocol"].(string); ok {
				server.Protocol = protocol
			}

			if desc, ok := serverMap["description"].(string); ok {
				server.Description = desc
			}

			result[name] = server
		}
	}

	return result
}

func parseChannels(channels map[string]any) map[string]Channel {
	result := make(map[string]Channel)

	for name, ch := range channels {
		if channelMap, ok := ch.(map[string]any); ok {
			channel := Channel{Extensions: make(map[string]any)}
			if desc, ok := channelMap["description"].(string); ok {
				channel.Description = desc
			}
			// Parse subscribe/publish operations (simplified)
			if sub, ok := channelMap["subscribe"].(map[string]any); ok {
				channel.Subscribe = parseOperation(sub)
			}

			if pub, ok := channelMap["publish"].(map[string]any); ok {
				channel.Publish = parseOperation(pub)
			}

			result[name] = channel
		}
	}

	return result
}

func parseAsyncComponents(components map[string]any) *AsyncComponents {
	result := &AsyncComponents{
		Messages:        make(map[string]map[string]any),
		Schemas:         make(map[string]map[string]any),
		SecuritySchemes: make(map[string]AsyncSecurityScheme),
	}

	if messages, ok := components["messages"].(map[string]any); ok {
		for name, msg := range messages {
			if msgMap, ok := msg.(map[string]any); ok {
				result.Messages[name] = msgMap
			}
		}
	}

	if schemas, ok := components["schemas"].(map[string]any); ok {
		for name, schema := range schemas {
			if schemaMap, ok := schema.(map[string]any); ok {
				result.Schemas[name] = schemaMap
			}
		}
	}

	// Parse security schemes
	if securitySchemes, ok := components["securitySchemes"].(map[string]any); ok {
		for name, scheme := range securitySchemes {
			if schemeMap, ok := scheme.(map[string]any); ok {
				sec := AsyncSecurityScheme{}
				if t, ok := schemeMap["type"].(string); ok {
					sec.Type = t
				}

				if desc, ok := schemeMap["description"].(string); ok {
					sec.Description = desc
				}

				if n, ok := schemeMap["name"].(string); ok {
					sec.Name = n
				}

				if in, ok := schemeMap["in"].(string); ok {
					sec.In = in
				}

				if s, ok := schemeMap["scheme"].(string); ok {
					sec.Scheme = s
				}

				if bf, ok := schemeMap["bearerFormat"].(string); ok {
					sec.BearerFormat = bf
				}

				if oidc, ok := schemeMap["openIdConnectUrl"].(string); ok {
					sec.OpenIdConnectURL = oidc
				}

				if flows, ok := schemeMap["flows"].(map[string]any); ok {
					sec.Flows = flows
				}

				result.SecuritySchemes[name] = sec
			}
		}
	}

	return result
}

func mergeChannels(existing, newChannel Channel) Channel {
	// Merge subscribe/publish operations
	if newChannel.Subscribe != nil {
		existing.Subscribe = newChannel.Subscribe
	}

	if newChannel.Publish != nil {
		existing.Publish = newChannel.Publish
	}

	return existing
}

// Helper functions for AsyncAPI composition

func shouldIncludeAsyncAPIInMerge(schema AsyncAPIServiceSchema) bool {
	for _, schemaDesc := range schema.Manifest.Schemas {
		if schemaDesc.Type == farp.SchemaTypeAsyncAPI &&
			schemaDesc.Metadata != nil &&
			schemaDesc.Metadata.AsyncAPI != nil {
			// Check for composition config when available
			return true
		}
	}

	return false
}

func getAsyncAPICompositionConfig(manifest *farp.SchemaManifest) *farp.CompositionConfig {
	for _, schemaDesc := range manifest.Schemas {
		if schemaDesc.Type == farp.SchemaTypeAsyncAPI &&
			schemaDesc.Metadata != nil &&
			schemaDesc.Metadata.AsyncAPI != nil {
			// For now, return nil - AsyncAPI doesn't have composition config yet
			// Could be added to spec later
			return nil
		}
	}

	return nil
}

func (m *AsyncAPIMerger) getConflictStrategy(config *farp.CompositionConfig) farp.ConflictStrategy {
	if config != nil && config.ConflictStrategy != "" {
		return config.ConflictStrategy
	}

	return m.config.DefaultConflictStrategy
}

func getAsyncAPIChannelPrefix(manifest *farp.SchemaManifest, config *farp.CompositionConfig) string {
	if config != nil && config.ComponentPrefix != "" {
		return config.ComponentPrefix
	}

	return manifest.ServiceName
}

func getAsyncAPIMessagePrefix(manifest *farp.SchemaManifest, config *farp.CompositionConfig) string {
	if config != nil && config.ComponentPrefix != "" {
		return config.ComponentPrefix
	}

	return manifest.ServiceName
}

// SortChannels sorts channels alphabetically.
func SortChannels(channels map[string]Channel) []string {
	keys := make([]string, 0, len(channels))
	for k := range channels {
		keys = append(keys, k)
	}

	sort.Strings(keys)

	return keys
}

File: farp-rust/src/merger/asyncapi.rs
//! AsyncAPI schema merger for combining multiple service schemas

use super::types::*;
use super::*;
use crate::errors::Result;
use crate::types::{SchemaManifest, SchemaType};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// AsyncAPI 2.x/3.x specification
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct AsyncAPISpec {
    pub asyncapi: String,
    pub info: Info,
    #[serde(skip_serializing_if = "HashMap::is_empty", default)]
    pub servers: HashMap<String, AsyncServer>,
    pub channels: HashMap<String, Channel>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub components: Option<AsyncComponents>,
    #[serde(skip_serializing_if = "Vec::is_empty", default)]
    pub security: Vec<HashMap<String, Vec<String>>>,
    #[serde(flatten)]
    pub extensions: HashMap<String, serde_json::Value>,
}

/// AsyncAPI server (broker connection)
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct AsyncServer {
    pub url: String,
    pub protocol: String, // kafka, amqp, mqtt, ws, etc.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub variables: Option<HashMap<String, serde_json::Value>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub bindings: Option<HashMap<String, serde_json::Value>>,
}

/// AsyncAPI channel
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Channel {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub subscribe: Option<Operation>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub publish: Option<Operation>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub parameters: Option<HashMap<String, serde_json::Value>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub bindings: Option<HashMap<String, serde_json::Value>>,
    #[serde(flatten)]
    pub extensions: HashMap<String, serde_json::Value>,
}

/// AsyncAPI components
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct AsyncComponents {
    #[serde(skip_serializing_if = "HashMap::is_empty", default)]
    pub messages: HashMap<String, serde_json::Value>,
    #[serde(skip_serializing_if = "HashMap::is_empty", default)]
    pub schemas: HashMap<String, serde_json::Value>,
    #[serde(
        skip_serializing_if = "HashMap::is_empty",
        default,
        rename = "securitySchemes"
    )]
    pub security_schemes: HashMap<String, AsyncSecurityScheme>,
}

/// AsyncAPI security scheme
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct AsyncSecurityScheme {
    #[serde(rename = "type")]
    pub scheme_type: String, // userPassword, apiKey, X509, etc.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "in")]
    pub in_: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub scheme: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "bearerFormat")]
    pub bearer_format: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "openIdConnectUrl")]
    pub openid_connect_url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub flows: Option<HashMap<String, serde_json::Value>>,
}

/// Service schema with AsyncAPI context
#[derive(Debug, Clone)]
pub struct AsyncAPIServiceSchema {
    pub manifest: SchemaManifest,
    pub schema: serde_json::Value,
    pub parsed: Option<AsyncAPISpec>,
}

/// AsyncAPI merger
pub struct AsyncAPIMerger {
    config: MergerConfig,
}

/// AsyncAPI merge result
#[derive(Debug, Clone)]
pub struct AsyncAPIMergeResult {
    pub spec: AsyncAPISpec,
    pub included_services: Vec<String>,
    pub excluded_services: Vec<String>,
    pub conflicts: Vec<Conflict>,
    pub warnings: Vec<String>,
}

impl AsyncAPIMerger {
    pub fn new(config: MergerConfig) -> Self {
        Self { config }
    }

    pub fn merge(&self, schemas: Vec<AsyncAPIServiceSchema>) -> Result<AsyncAPIMergeResult> {
        let mut result = AsyncAPIMergeResult {
            spec: AsyncAPISpec {
                asyncapi: "2.6.0".to_string(),
                info: Info {
                    title: self.config.merged_title.clone(),
                    description: Some(self.config.merged_description.clone()),
                    version: self.config.merged_version.clone(),
                    terms_of_service: None,
                    contact: None,
                    license: None,
                    extensions: HashMap::new(),
                },
                servers: HashMap::new(),
                channels: HashMap::new(),
                components: Some(AsyncComponents {
                    messages: HashMap::new(),
                    schemas: HashMap::new(),
                    security_schemes: HashMap::new(),
                }),
                security: Vec::new(),
                extensions: HashMap::new(),
            },
            included_services: Vec::new(),
            excluded_services: Vec::new(),
            conflicts: Vec::new(),
            warnings: Vec::new(),
        };

        let mut seen_channels: HashMap<String, String> = HashMap::new();
        let mut seen_messages: HashMap<String, String> = HashMap::new();
        let mut seen_servers: HashMap<String, String> = HashMap::new();
        let mut seen_security_schemes: HashMap<String, String> = HashMap::new();

        for mut schema in schemas {
            let service_name = schema.manifest.service_name.clone();

            if !should_include_asyncapi(&schema) {
                result.excluded_services.push(service_name);
                continue;
            }

            result.included_services.push(service_name.clone());

            if schema.parsed.is_none() {
                match parse_asyncapi_schema(&schema.schema) {
                    Ok(parsed) => schema.parsed = Some(parsed),
                    Err(e) => {
                        result.warnings.push(format!(
                            "Failed to parse AsyncAPI schema for {service_name}: {e}"
                        ));
                        continue;
                    }
                }
            }

            let parsed = schema.parsed.as_ref().unwrap();
            let strategy = self.config.default_conflict_strategy;

            let channel_prefix = &schema.manifest.service_name;
            let message_prefix = &schema.manifest.service_name;

            // Merge channels
            for (channel_name, channel) in &parsed.channels {
                let mut prefixed_name = format!("{channel_prefix}.{channel_name}");

                if let Some(existing_service) = seen_channels.get(&prefixed_name) {
                    let conflict = Conflict {
                        conflict_type: ConflictType::Component,
                        item: channel_name.clone(),
                        services: vec![existing_service.clone(), service_name.clone()],
                        resolution: String::new(),
                        strategy,
                    };

                    match strategy {
                        ConflictStrategy::Error => {
                            return Err(crate::errors::Error::Custom(format!(
                                "channel conflict: {channel_name} exists in both {existing_service} and {service_name}"
                            )));
                        }
                        ConflictStrategy::Skip => {
                            let mut c = conflict;
                            c.resolution = format!("Skipped channel from {service_name}");
                            result.conflicts.push(c);
                            continue;
                        }
                        ConflictStrategy::Overwrite => {
                            let mut c = conflict;
                            c.resolution = format!("Overwritten with {service_name} version");
                            result.conflicts.push(c);
                        }
                        ConflictStrategy::Prefix => {
                            prefixed_name = format!("{service_name}.{channel_name}");
                            let mut c = conflict;
                            c.resolution = format!("Prefixed to {prefixed_name}");
                            result.conflicts.push(c);
                        }
                        ConflictStrategy::Merge => {
                            let existing = result.spec.channels.get(&prefixed_name).cloned();
                            if let Some(existing) = existing {
                                let merged = merge_channels(existing, channel.clone());
                                result.spec.channels.insert(prefixed_name.clone(), merged);
                            }
                            let mut c = conflict;
                            c.resolution = "Merged operations".to_string();
                            result.conflicts.push(c);
                            continue;
                        }
                    }
                }

                result
                    .spec
                    .channels
                    .insert(prefixed_name.clone(), channel.clone());
                seen_channels.insert(prefixed_name, service_name.clone());
            }

            // Merge components
            if let Some(components) = &parsed.components {
                // Merge messages
                for (name, message) in &components.messages {
                    let prefixed_name = format!("{message_prefix}_{name}");
                    if let Some(existing_service) = seen_messages.get(&prefixed_name) {
                        if strategy == ConflictStrategy::Skip {
                            result.conflicts.push(Conflict {
                                conflict_type: ConflictType::Component,
                                item: name.clone(),
                                services: vec![existing_service.clone(), service_name.clone()],
                                resolution: format!("Skipped message from {service_name}"),
                                strategy,
                            });
                            continue;
                        }
                    }

                    if let Some(spec_components) = result.spec.components.as_mut() {
                        spec_components
                            .messages
                            .insert(prefixed_name.clone(), message.clone());
                    }
                    seen_messages.insert(prefixed_name, service_name.clone());
                }

                // Merge schemas
                for (name, schema_obj) in &components.schemas {
                    let prefixed_name = format!("{message_prefix}_{name}");
                    if let Some(spec_components) = result.spec.components.as_mut() {
                        spec_components
                            .schemas
                            .insert(prefixed_name, schema_obj.clone());
                    }
                }

                // Merge security schemes
                for (name, sec_scheme) in &components.security_schemes {
                    if let Some(existing_service) = seen_security_schemes.get(name) {
                        let conflict = Conflict {
                            conflict_type: ConflictType::SecurityScheme,
                            item: name.clone(),
                            services: vec![existing_service.clone(), service_name.clone()],
                            resolution: String::new(),
                            strategy,
                        };

                        match strategy {
                            ConflictStrategy::Error => {
                                return Err(crate::errors::Error::Custom(format!(
                                    "security scheme conflict: {name} exists in both {existing_service} and {service_name}"
                                )));
                            }
                            ConflictStrategy::Skip => {
                                let mut c = conflict;
                                c.resolution =
                                    format!("Skipped security scheme from {service_name}");
                                result.conflicts.push(c);
                                continue;
                            }
                            ConflictStrategy::Overwrite => {
                                let mut c = conflict;
                                c.resolution = format!("Overwritten with {service_name} version");
                                result.conflicts.push(c);
                            }
                            ConflictStrategy::Prefix => {
                                let prefixed_name = format!("{service_name}_{name}");
                                let mut c = conflict;
                                c.resolution = format!("Prefixed to {prefixed_name}");
                                result.conflicts.push(c);
                                if let Some(spec_components) = result.spec.components.as_mut() {
                                    spec_components
                                        .security_schemes
                                        .insert(prefixed_name.clone(), sec_scheme.clone());
                                }
                                seen_security_schemes.insert(prefixed_name, service_name.clone());
                                continue;
                            }
                            ConflictStrategy::Merge => {
                                let mut c = conflict;
                                c.resolution =
                                    format!("Merged (overwritten) with {service_name} version");
                                result.conflicts.push(c);
                            }
                        }
                    }

                    if let Some(spec_components) = result.spec.components.as_mut() {
                        spec_components
                            .security_schemes
                            .insert(name.clone(), sec_scheme.clone());
                    }
                    seen_security_schemes.insert(name.clone(), service_name.clone());
                }
            }

            // Merge servers
            for (server_name, server) in &parsed.servers {
                let prefixed_name = format!("{service_name}_{server_name}");
                if let Some(existing_service) = seen_servers.get(&prefixed_name) {
                    result.warnings.push(format!(
                        "Server {server_name} from {service_name} overwrites {existing_service}"
                    ));
                }
                result
                    .spec
                    .servers
                    .insert(prefixed_name.clone(), server.clone());
                seen_servers.insert(prefixed_name, service_name.clone());
            }
        }

        Ok(result)
    }
}

/// Parse AsyncAPI schema from JSON
pub fn parse_asyncapi_schema(raw: &serde_json::Value) -> Result<AsyncAPISpec> {
    let schema_map = raw
        .as_object()
        .ok_or_else(|| crate::errors::Error::invalid_schema("schema must be an object"))?;

    let asyncapi = schema_map
        .get("asyncapi")
        .and_then(|v| v.as_str())
        .ok_or_else(|| crate::errors::Error::invalid_schema("missing asyncapi version"))?
        .to_string();

    let info = super::openapi::parse_info_public(schema_map.get("info"))?;

    let servers = schema_map
        .get("servers")
        .and_then(|v| v.as_object())
        .map(parse_async_servers)
        .unwrap_or_default();

    let channels = schema_map
        .get("channels")
        .and_then(|v| v.as_object())
        .map(parse_channels)
        .unwrap_or_default();

    let components = schema_map
        .get("components")
        .and_then(|v| v.as_object())
        .map(parse_async_components);

    Ok(AsyncAPISpec {
        asyncapi,
        info,
        servers,
        channels,
        components,
        security: Vec::new(),
        extensions: HashMap::new(),
    })
}

fn parse_async_servers(
    obj: &serde_json::Map<String, serde_json::Value>,
) -> HashMap<String, AsyncServer> {
    obj.iter()
        .filter_map(|(name, server)| {
            server.as_object().and_then(|s| {
                Some((
                    name.clone(),
                    AsyncServer {
                        url: s.get("url")?.as_str()?.to_string(),
                        protocol: s.get("protocol")?.as_str()?.to_string(),
                        description: s
                            .get("description")
                            .and_then(|v| v.as_str())
                            .map(String::from),
                        variables: None,
                        bindings: None,
                    },
                ))
            })
        })
        .collect()
}

fn parse_channels(obj: &serde_json::Map<String, serde_json::Value>) -> HashMap<String, Channel> {
    obj.iter()
        .filter_map(|(name, channel)| {
            channel.as_object().map(|c| {
                (
                    name.clone(),
                    Channel {
                        description: c
                            .get("description")
                            .and_then(|v| v.as_str())
                            .map(String::from),
                        subscribe: c
                            .get("subscribe")
                            .and_then(|v| v.as_object())
                            .map(super::openapi::parse_operation_public),
                        publish: c
                            .get("publish")
                            .and_then(|v| v.as_object())
                            .map(super::openapi::parse_operation_public),
                        parameters: None,
                        bindings: None,
                        extensions: HashMap::new(),
                    },
                )
            })
        })
        .collect()
}

fn parse_async_components(obj: &serde_json::Map<String, serde_json::Value>) -> AsyncComponents {
    AsyncComponents {
        messages: obj
            .get("messages")
            .and_then(|v| v.as_object())
            .map(|m| m.iter().map(|(k, v)| (k.clone(), v.clone())).collect())
            .unwrap_or_default(),
        schemas: obj
            .get("schemas")
            .and_then(|v| v.as_object())
            .map(|s| s.iter().map(|(k, v)| (k.clone(), v.clone())).collect())
            .unwrap_or_default(),
        security_schemes: HashMap::new(),
    }
}

fn merge_channels(existing: Channel, new: Channel) -> Channel {
    Channel {
        description: new.description.or(existing.description),
        subscribe: new.subscribe.or(existing.subscribe),
        publish: new.publish.or(existing.publish),
        parameters: new.parameters.or(existing.parameters),
        bindings: new.bindings.or(existing.bindings),
        extensions: {
            let mut ext = existing.extensions;
            ext.extend(new.extensions);
            ext
        },
    }
}

fn should_include_asyncapi(schema: &AsyncAPIServiceSchema) -> bool {
    schema
        .manifest
        .schemas
        .iter()
        .any(|s| s.schema_type == SchemaType::AsyncAPI)
}
